### PR TITLE
[compiler-v2] Externalize the linter architecture

### PR DESCRIPTION
**Pull request by vineethk**
## Description

The goal of this PR is to make the compiler v2 accept an option to run externally provided lint checks. The compiler provides necessary traits for external implementors of the lint checks (on AST and stackless bytecode).

Further, all the lint checks inside the compiler are now moved out to a new (third-party) crate called `move-linter`, which is expected to contain aptos move specific lint checks. The corresponding tests have also been moved.

The aptos CLI lint tool (`aptos move lint`) invokes the compiler v2 with the checks inside `move-linter` crate: thus, the existing behavior is unchanged. In the future, it could implement some of the extended checks and even additional checks (such as those particular to aptos framework and ecosystem) and pass along those checks as well, in addition to the checks in `move-linter`.

A follow up PR will introduce `lint.toml` to be able to configure the linter, which would be passed to the linter factory making up the lint check instances.

## Other implementation options considered

- Making the external checks part of various configs/options like `CompilerConfig`, `BuildOptions` etc. to further minimize API changes. However, this requires implementing serialization/deserialization, `PartialEq` and other traits for dynamic trait objects.
- Replicating the `build` behavior (dependency resolution, compiler driver, etc) for including external checks - this results in a lot of code duplication, which could hamper package system revamping.

After attempting both these options, I settled on changing a handful of APIs (only those needed, we could expand to more on a need basis) to explicitly pass external checks where necessary.

## How Has This Been Tested?
- Existing tests (lint related checks moved from compiler v2 to `move-linter` crate)
- Manually tested the `aptos move lint` CLI command.

Generally speaking, none of the user-visible behavior should have changed from this PR.

## Type of Change
- [x] Refactoring

## Which Components or Systems Does This Change Impact?
- [x] Aptos CLI/SDK
- [x] Move Compiler

### CODE REVIEW

**Comment by vineethk** [comment]
* **#15138** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/15138?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @vineethk and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by rahxephon89 — APPROVED** [review]
LGTM, just left a minor comment

File: third_party/move/tools/move-package/tests/test_runner.rs [hunk 1]
```diff
@@ -81,9 +81,9 @@ fn run_test_impl(
                 )
                 .into())
             },
-            (true, _) => match BuildPlan::create(resolved_package)
-                .and_then(|bp| bp.compile_no_exit(&compiler_config.clone(), &mut Vec::new()))
-            {
+            (true, _) => match BuildPlan::create(resolved_package).and_then(|bp| {
+                bp.compile_no_exit(&compiler_config.clone(), vec![], &mut Vec::new())
```
Comment (brmataptos) [inline third_party/move/tools/move-package/tests/test_runner.rs:85]: The juxtaposition of two different ways to create an empty vector is kind of odd.  How about
```
.., vec![], &mut vec![])
```

**Review by brmataptos — APPROVED** [review]
Looks reasonable, see a few minor comments.

**Review by wrwg — APPROVED** [review]

**Review by brmataptos — APPROVED** [review]

File: third_party/move/tools/move-linter/tests/stackless_bytecode_lints/needless_mutable_reference_warn.exp [hunk 2]
```diff
@@ -1,150 +1,143 @@
...
-    ┌─ tests/lints/stackless_bytecode_lints/needless_mutable_reference_warn.move:212:9
+    ┌─ tests/stackless_bytecode_lints/needless_mutable_reference_warn.move:212:9
     │
 212 │         *&mut u
     │         ^^^^^^^
     │
     = To suppress this warning, annotate the function/module with the attribute `#[lint::skip(needless_deref_ref)]`.
 
-
-Diagnostics:
```
[the remark this answers is not in the document]
Comment (vineethk) [reply third_party/move/tools/move-linter/tests/stackless_bytecode_lints/needless_mutable_reference_warn.exp:28]: No, we are not suppressing all normal compiler warnings here. This warning just ended up moving elsewhere in the expected file, but is still present. The difference is due to the way compiler-v2's testsuite prints out warnings vs how the compiler API prints out the warnings (the lint crate uses the compiler API).

There are already some tests with compiler errors here: e.g., `third_party/move/tools/move-linter/tests/model_ast_lints/bad_lint_attribute_01.exp.`

File: third_party/move/move-compiler-v2/src/external_checks.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    fn get_stackless_bytecode_checkers(&self) -> Vec<Box<dyn StacklessBytecodeChecker>>;
+}
+
+impl fmt::Debug for dyn ExternalChecks {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let exp_checkers = self
+            .get_exp_checkers()
+            .into_iter()
+            .map(|c| c.get_name())
+            .collect::<Vec<_>>()
```
[the remark this answers is not in the document]
Comment (brmataptos) [reply third_party/move/move-compiler-v2/src/external_checks.rs:30]: https://www.reddit.com/r/rust/comments/6q4uqc/help_whats_the_best_way_to_join_an_iterator_of/, maybe it's in itertools so not visible by default.

File: crates/aptos/src/move_tool/lint.rs [hunk 4]
```diff
@@ -112,7 +113,7 @@ impl CliCommand<&'static str> for LintPackage {
 
     async fn execute(self) -> CliTypedResult<&'static str> {
         let move_options = MovePackageDir {
-            compiler_version: Some(CompilerVersion::V2_0),
+            compiler_version: Some(CompilerVersion::latest_stable()),
```
[the remark this answers is not in the document]
Comment (brmataptos) [reply crates/aptos/src/move_tool/lint.rs:116]: FYI, I filed an issues  https://github.com/aptos-labs/aptos-core/issues/15178 to track this.

**Review by brmataptos — APPROVED** [review]

**Pull request merged**